File: src/main/java/fun/spmc/smpmod/registry/QuestRegistry.java
package fun.spmc.smpmod.registry;

import fun.spmc.smpmod.quest.QuestManager;
import fun.spmc.smpmod.quest.data.Quest;
import net.fabricmc.fabric.api.entity.event.v1.ServerLivingEntityEvents;
import net.fabricmc.fabric.api.event.player.PlayerBlockBreakEvents;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.Identifier;
import net.minecraft.server.level.ServerPlayer;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class QuestRegistry {
    private static final Map<String, Quest> QUESTS = new HashMap<>();

    public static void init() {
        initDaily();
        initWeekly();

        // demo npc quests
        /**
        register(new Quest(
                "blacksmith_1", "Gathering Supplies", "Bring 10 Coal to the Blacksmith",
                QuestCategory.NPC,
                QuestType.GATHER_ITEM,
                Identifier.withDefaultNamespace("coal"),
                10,
                new QuestReward(50.0, 20, List.of(new ItemStack(Items.IRON_INGOT, 3))),
                Optional.of("blacksmith"),
                Optional.empty()
        ));

        register(new Quest(
                "blacksmith_2", "First Blade", "Craft an Iron Sword",
                QuestCategory.NPC,
                QuestType.CRAFTING,
                Identifier.withDefaultNamespace("iron_sword"),
                1,
                new QuestReward(200.0, 100, List.of(new ItemStack(Items.DIAMOND, 1))),
                Optional.of("blacksmith"),
                Optional.of("blacksmith_1")
        ));*/

        PlayerBlockBreakEvents.AFTER.register((_, player, _, state, _) -> QuestManager.getQuests((ServerPlayer) player).getActiveQuests().forEach(activeQuest -> {
            Quest quest = activeQuest.getQuest();
            if (quest.type() == Quest.QuestType.MINE_BLOCK && quest.target().equals(BuiltInRegistries.BLOCK.getKey(state.getBlock()))) activeQuest.increment(1);
        }));

        ServerLivingEntityEvents.AFTER_DEATH.register((entity, damageSource) -> {
            if (damageSource.getEntity() instanceof ServerPlayer player) {
                Identifier mobId = BuiltInRegistries.ENTITY_TYPE.getKey(entity.getType());

                QuestManager.getQuests(player).getActiveQuests().forEach(activeQuest -> {
                    Quest quest = activeQuest.getQuest();
                    if (quest.type() == Quest.QuestType.KILL_MOB && quest.target().equals(mobId)) {
                        activeQuest.increment(1);
                    }
                });
            }
        });
    }

    private static void initDaily() {
        register(new Quest(
                "daily_mine_iron", "Iron Miner", "Mine 16 Iron Ore",
                Quest.QuestType.MINE_BLOCK, Identifier.withDefaultNamespace("iron_ore"), 16, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(3575, 75, List.of()), Optional.empty(), Optional.empty()
        ));

        register(new Quest(
                "daily_mine_stone", "Stone Miner", "Mine 64 Stone",
                Quest.QuestType.MINE_BLOCK, Identifier.withDefaultNamespace("stone"), 64, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(2500, 80, List.of()), Optional.empty(), Optional.empty()
        ));

        register(new Quest(
                "daily_kill_zombie", "Zombie Killer", "Kill 15 Zombies",
                Quest.QuestType.KILL_MOB, Identifier.withDefaultNamespace("zombie"), 15, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(4555, 120, List.of()), Optional.empty(), Optional.empty()
        ));

        register(new Quest(
                "daily_kill_creeper", "Creeper Killer", "Kill 15 Creepers",
                Quest.QuestType.KILL_MOB, Identifier.withDefaultNamespace("creeper"), 15, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(4555, 120, List.of()), Optional.empty(), Optional.empty()
        ));

        register(new Quest(
                "daily_kill_skeleton", "Skeleton Killer", "Kill 15 Skeletons",
                Quest.QuestType.KILL_MOB, Identifier.withDefaultNamespace("skeleton"), 15, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(4555, 120, List.of()), Optional.empty(), Optional.empty()
        ));

        register(new Quest(
                "daily_fish_1", "Fishing Newbie", "Fish 15 Times",
                Quest.QuestType.FISHING, Identifier.withDefaultNamespace("fishing"), 15, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(1550, 45, List.of()), Optional.empty(), Optional.empty()
        ));

        register(new Quest(
                "daily_fish_2", "Fishing Amateur", "Fish 35 Times",
                Quest.QuestType.FISHING, Identifier.withDefaultNamespace("fishing"), 35, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(3750, 65, List.of()), Optional.empty(), Optional.empty()
        ));

        register(new Quest(
                "daily_fish_3", "Fishing Master", "Fish 75 Times",
                Quest.QuestType.FISHING, Identifier.withDefaultNamespace("fishing"), 75, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(5900, 85, List.of()), Optional.empty(), Optional.empty()
        ));
    }

    private static void initWeekly() {
        register(new Quest(
                "daily_fish_3", "Fishing Master", "Fish 75 Times",
                Quest.QuestType.FISHING, Identifier.withDefaultNamespace("fishing"), 75, Quest.QuestCategory.DAILY,
                new Quest.QuestReward(5900, 85, List.of()), Optional.empty(), Optional.empty()
        ));
    }

    private static void register(Quest quest) { QUESTS.put(quest.id(), quest); }
    public static Quest get(String id) { return QUESTS.get(id); }
    public static List<Quest> getAllForWeekly() { return QUESTS.values().stream().filter(q -> q.questType().equals(Quest.QuestCategory.WEEKLY)).toList(); }
    public static List<Quest> getAllForDaily() { return QUESTS.values().stream().filter(q -> q.questType().equals(Quest.QuestCategory.DAILY)).toList(); }
    public static List<Quest> getAllForNpc(String npcId) { return QUESTS.values().stream().filter(q -> q.npcId().map(id -> id.equals(npcId)).orElse(false)).toList(); }
}

File: src/main/java/spmc/smpmod/registry/QuestRegistry.kt
package spmc.smpmod.registry

import net.fabricmc.fabric.api.entity.event.v1.ServerLivingEntityEvents
import net.fabricmc.fabric.api.event.player.PlayerBlockBreakEvents
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.server.level.ServerPlayer
import spmc.smpmod.quest.Quest
import spmc.smpmod.quest.Quest.QuestReward
import spmc.smpmod.quest.QuestManager.Companion.getQuests
import spmc.smpmod.utils.anyInCreative

object QuestRegistry {
	private val QUESTS: MutableMap<String, Quest> = mutableMapOf()

	fun init() {
		initDaily()
		initWeekly()

		PlayerBlockBreakEvents.AFTER.register { _, player, _, state, _ ->
			if (anyInCreative(player as? ServerPlayer ?: return@register)) return@register
			getQuests(player).activeQuests.forEach { val quest = it.getQuest() ?: return@register
				if (quest.type == Quest.QuestType.MINE_BLOCK && quest.target == BuiltInRegistries.BLOCK.getKey(state.block)) it.increment(1)
			}
		}

		ServerLivingEntityEvents.AFTER_DEATH.register { entity, damageSource ->
			if (entity.level().dimension().identifier().namespace != "minecraft") return@register
			getQuests(damageSource.entity as? ServerPlayer?: return@register).activeQuests.forEach { val quest = it.getQuest() ?: return@register
				if (quest.type == Quest.QuestType.KILL_MOB && quest.target == BuiltInRegistries.ENTITY_TYPE.getKey(entity.type)) it.increment(1)
			}
		}
	}

	private fun initDaily() {
		register(Quest("dm_iro", "Iron Miner", Quest.QuestType.MINE_BLOCK, "iron_ore", 16, Quest.QuestCategory.DAILY, QuestReward(950, 75)))
		register(Quest("dm_dia", "Diamond Miner", Quest.QuestType.MINE_BLOCK, "diamond_block", 16, Quest.QuestCategory.DAILY, QuestReward(1200, 75)))
		register(Quest("dm_sto", "Stone Miner", Quest.QuestType.MINE_BLOCK, "stone", 64, Quest.QuestCategory.DAILY, QuestReward(675, 80)))
		register(Quest("dk_zom", "Zombie Killer", Quest.QuestType.KILL_MOB, "zombie", 15, Quest.QuestCategory.DAILY, QuestReward(675, 120)))
		register(Quest("dk_cre", "Creeper Killer", Quest.QuestType.KILL_MOB, "creeper", 15, Quest.QuestCategory.DAILY, QuestReward(639, 120)))
		register(Quest("dk_ske", "Skeleton Killer", Quest.QuestType.KILL_MOB, "skeleton", 15, Quest.QuestCategory.DAILY, QuestReward(779, 120)))
		register(Quest("df1", "Fishing Newbie", Quest.QuestType.FISHING, "fishing", 15, Quest.QuestCategory.DAILY, QuestReward(173, 45)))
		register(Quest("df2", "Fishing Amateur", Quest.QuestType.FISHING, "fishing", 35, Quest.QuestCategory.DAILY, QuestReward(570, 65)))
		register(Quest("df3", "Fishing Master", Quest.QuestType.FISHING, "fishing", 75, Quest.QuestCategory.DAILY, QuestReward(1000, 85)))
	}

	private fun initWeekly() {
		register(Quest("wm_obs", "Obsidian Collector", Quest.QuestType.MINE_BLOCK, "obsidian", 384, Quest.QuestCategory.WEEKLY, QuestReward(6000, 150)))
		register(Quest("wm_net", "Ancient Archaeologist", Quest.QuestType.MINE_BLOCK, "ancient_debris", 192, Quest.QuestCategory.WEEKLY, QuestReward(6500, 160)))
		register(Quest("wk_war", "Warden Harvester", Quest.QuestType.KILL_MOB, "warden", 15, Quest.QuestCategory.WEEKLY, QuestReward(4600, 220)))
		register(Quest("wk_wit", "Wither Hunter", Quest.QuestType.KILL_MOB, "wither", 3, Quest.QuestCategory.WEEKLY, QuestReward(5700, 240)))
		register(Quest("wk_dra", "End Guardian", Quest.QuestType.KILL_MOB, "ender_dragon", 2, Quest.QuestCategory.WEEKLY, QuestReward(6200, 375)))
		register(Quest("wk_elg", "Elder Reaper", Quest.QuestType.KILL_MOB, "elder_guardian", 15, Quest.QuestCategory.WEEKLY, QuestReward(7500, 400)))
		register(Quest("wf1", "Fishing Ascendant", Quest.QuestType.FISHING, "fishing", 150, Quest.QuestCategory.WEEKLY, QuestReward(2250, 115)))
		register(Quest("wf2", "Fishing Grandmaster", Quest.QuestType.FISHING, "fishing", 350, Quest.QuestCategory.WEEKLY, QuestReward(4750, 150)))
		register(Quest("wf3", "Fishing Deity", Quest.QuestType.FISHING, "fishing", 750, Quest.QuestCategory.WEEKLY, QuestReward(6800, 175)))
	}

	val allForWeekly = QUESTS.values.filter { it.questType == Quest.QuestCategory.WEEKLY }
	val allForDaily = QUESTS.values.filter { it.questType == Quest.QuestCategory.DAILY }

	private fun register(quest: Quest) { QUESTS[quest.id] = quest }
	fun get(id: String): Quest? = QUESTS[id]
	fun getAllForNpc(npcId: String): List<Quest?> = QUESTS.values.filter { q -> q.npcId.map { it == npcId }.orElse(false)!! }
}